mm2src/coins/z_coin.rs

```diff
@@ -1,94 +1,564 @@
+                &ARRRConsensusParams {},
+                BlockHeight::from_u32(height as u32),
+                &z_cash_tx,
+                &ext,
+            );
+            println!("Decrypted len {}", decrypted.len());
+            let decrypted_output = decrypted
+                .iter()
+                .find(|out| out.index as u32 == unspent.out_index)
+                .or_mm_err(|| GenTxError::DecryptedOutputNotFound)?;
+            let witness = self.get_unspent_witness(&decrypted_output.note, height as u32).await?;
+            selected_notes_with_witness.push((decrypted_output.note.clone(), witness));
+        }
+
+        for (note, witness) in selected_notes_with_witness {
+            tx_builder.add_sapling_spend(
+                self.z_fields.z_spending_key.clone(),
+                *self.z_fields.my_z_addr.diversifier(),
+                note,
+                witness.path().or_mm_err(|| GenTxError::FailedToGetMerklePath)?,
+            )?;
+        }
+
+        for z_out in z_outputs {
+            if z_out.to_addr == self.z_fields.my_z_addr {
+                received_by_me += u64::from(z_out.amount);
+            }
+
+            tx_builder.add_sapling_output(z_out.viewing_key, z_out.to_addr, z_out.amount, z_out.memo)?;
+        }
+
+        if change > BigDecimal::from(0) {
+            let change_sat = sat_from_big_decimal(&change, self.utxo_arc.decimals)?;
+            received_by_me += change_sat;
+
+            tx_builder.add_sapling_output(
+                None,
+                self.z_fields.my_z_addr.clone(),
+                Amount::from_u64(change_sat).map_to_mm(|_| {
+                    GenTxError::NumConversion(NumConversError(format!(
+                        "Failed to get ZCash amount from {}",
+                        change_sat
+                    )))
+                })?,
+                None,
+            )?;
+        }
+
+        for output in t_outputs {
+            tx_builder.add_tx_out(output);
+        }
+
+        let (tx, _) = tx_builder.build(consensus::BranchId::Sapling, &self.z_fields.z_tx_prover)?;
+
+        let additional_data = AdditionalTxData {
+            received_by_me,
+            spent_by_me: sat_from_big_decimal(&total_input_amount, self.decimals())?,
+            fee_amount: sat_from_big_decimal(&tx_fee, self.decimals())?,
+            unused_change: None,
+            kmd_rewards: None,
+        };
+        Ok((tx, additional_data))
+    }
+
+    pub async fn send_outputs(
+        &self,
+        t_outputs: Vec<TxOut>,
+        z_outputs: Vec<ZOutput>,
+    ) -> Result<ZTransaction, MmError<SendOutputsErr>> {
+        let (tx, _) = self.gen_tx(t_outputs, z_outputs).await?;
+        let mut tx_bytes = Vec::with_capacity(1024);
+        tx.write(&mut tx_bytes).expect("Write should not fail");
+
+        self.rpc_client().send_raw_transaction(tx_bytes.into()).compat().await?;
+
+        self.rpc_client()
+            .wait_for_confirmations(
```

Comment:
Due to Sapling specifics, the shielded output becomes spendable only after it's confirmed. So we have to wait for tx to be mined to allow construction of a new one. I've realized that `z_unspent_mutex` should be locked during this process.

Reply:
I'd suggest adding the sent transaction to a container inside `BCH` coin. So when any other SLP token tries to generate a new transaction, it will wait until every transaction from this container is mined.

Reply:
Excuse me, did you intend to post this comment to another place? `ZCoin` has little relation to BCH and SLP.